File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.js
function solution(numbers) {
    const digit = numbers.split('');
    const visited = new Array(digit.length).fill(false);
    const numSet = new Set();
    
    function dfs(path){
        if(path.length > 0){
            const num = parseInt(path.join(''), 10);
            numSet.add(num);
        }
        
        for(let i = 0; i < digit.length; i++){
            if(visited[i]) continue;
            
            visited[i] = true;
            path.push(digit[i]);
            
            dfs(path);
            
            visited[i] = false;
            path.pop();
        }
    }
    
    dfs([]);
    
    function isPrime(num){
        if(num < 2) return false;
        for(let i = 2; i <= Math.sqrt(num); i++){
            if(num % i === 0){
                return false;
            }
        }
        return true;
    }
    
    let answer = 0;
    numSet.forEach(num => {
        if(isPrime(num)) answer++;
    });
    
    return answer;
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;
class Solution {
    static int answer = 0;
    static boolean[] visited;
    static Set<Integer> set = new HashSet<>();
    
    public int solution(String numbers) {
        visited = new boolean[numbers.length()];
        permu(numbers, "");
        for(int s : set){
            if(isPrime(s)){
                answer++;
            }
        }
        return answer;
    }
    
    public void permu(String numbers, String current){
        if(current.length() != 0){
            set.add(Integer.parseInt(current));
        }
        for(int i = 0; i < numbers.length(); i++){
            if(!visited[i]){
                visited[i] = true;
                permu(numbers, current + numbers.charAt(i));
                visited[i] = false;
            }   
        }
    }
    
    public boolean isPrime(int val){
        if (val < 2) return false;
        
        for(int i = 2; i <= Math.sqrt(val); i++){
            if(val % i == 0){
                return false;
            }
        }
        return true;
    }
}